### DISCUSSION TITLE
Need help transitioning from `tailwindcss-react-native`

### DISCUSSION THREAD

**Discussion by matallui** [nativewind/nativewind#194] [Q&A]
We have this Expo project (currently on SDK 46 with `react-native-web` 0.18.7, and everything was working fine for both native and web when using `tailwindcss-react-native`.

My goal is to try to switch to `nativewind` and, for that, this is what I did:

 - `yarn remove tailwindcss-react-native`
 - `yarn remove tailwindcss`
 - `yarn add nativewind`
 - `yarn add -D tailwindcss`
 - remove `TailwindProvider` from `App.tsx`
 - update `babel.config.js` to use `nativewind/babel`
 - update `nativewind.d.ts` (for TS support)
 
 After doing that, I'm having issues when I try to start the app (both native and web).
 
 I get some exception saying:
 
```
Unhandled JS Exception: undefined is not an object (evaluating '_NativeWindStyleSheet.create')
```

On web the errors are different, but at least I want to fix native first.

What could I be missing?

**Comment by marklawlor** [comment]
Do you have a link to your repo? It could be similar to this issue https://github.com/marklawlor/nativewind/issues/191 which was using a `tailwind.config.js` with an invalid file name.

**Reply by matallui** [reply]
Unfortunately it's a private repo, but I've managed to get it working.
I created a clean expo app and tried it and it worked, so I figured I wasn't doing anything wrong.

All I had to do to fix it was:

```
git clean -xfd .
git co .
yarn
npx expo start
```

After cleaning up everything and rebuilding it seemed to work.

When I tried the web though, I saw the styles were not fully working.
However, after I added the following at the top of my `App.tsx`, it started working again:

```tsx
import { NativeWindStyleSheet } from "nativewind";

NativeWindStyleSheet.setOutput({
  default: "native",
});
```

Is this the correct way to fix things for web? The docs say you need this for SDK <= 45 but I'm running SDK 46. Not sure if the same still applies.

**Reply by marklawlor** [reply]
NativeWind changes the default styling to use CSS for web https://www.nativewind.dev/guides/tailwindcss-react-native#css-is-the-default-for-web

I recommend editing your webpack config to include Tailwind CSS. How you do this depends on your project setup, but there is an example here https://github.com/marklawlor/nativewind/tree/next/examples/expo

**Reply by matallui** [reply]
Actually, just tried to run this on native and still have the original error I mentioned before, so not sure I've fixed everything yet.

Yeah, I saw that example but it seemed like I had to add a lot of dependencies (`postcss` & `postcss-loader`) and then add files I didn't have before (`main.css`, `postcss.config.js`, `webpack.config.js`), so it looked like a lot of changes to get it working, since this was all working fine with `tailwindcss-react-native`.

It was not clear to me exactly what I had to do to get it setup for web, even though there's a web section in the docs.

**Reply by matallui** [reply]
Btw, my project setup is pretty much exactly what you'd get if you ran:
```
yarn create expo-app -t expo-template-blank-typescript
```

**Reply by matallui** [reply]
Ok, I got over the issue by running `npx expo start -c`. Not sure why the bundler cache was not clear with `git clean -xfd .` though.

On a different note, I've noticed some styles have changed. The shadows are darker and standout much more than they did with `tailwindcss-react-native`. Not sure if this is normal.

**Reply by marklawlor** [reply]
While running on web, NativeWind acts as a compatibility layer between React Native and Tailwind CSS, you still need to setup Tailwind CSS yourself.

Expo Web is just a React & webpack project, so you will need to add Tailwind CSS as if you were a normal web app using react & webpack. If you were using Expo with Next.js/Gatsby you would follow the Tailwind CSS setup for those frameworks.

Hope that makes sense, if you have any suggestions on how I can improve the documentation I would appreciate it.

You can switch back to using React Native styles on web, but that will affect performance, break SSR and reduce the number of classes you can use.

**Reply by matallui** [reply]
Oh, I see... So I guess thus far I've only been using the `native` version of things by putting that snippet above in `App.tsx`.
I will probably need to understand how to add better web support to the Expo app and use CSS instead.
For now I'll use `native` so it doesn't break what we had done, since my goal is to simply transition from `tailwindcss-react-native`.

**Reply by matallui** [reply]
Is it normal that the shadows are all quite different on native though?

**Reply by marklawlor** [reply]
The shadows were quite buggy with `tailwindcss-react-native` so they were updated. They are more aligned now with the other React Native tailwind libraries for people migrating across. Sorry if that affected you, can adjust the values in your theme to return to the previous values.

**Reply by matallui** [reply]
No, that's fine! Our app is still in a PoC stage, so we're just playing around with things.

I just wanted to make sure the shadow differences were normal and that you were aware of it. I will fix the styling no problem. Tailwind makes it super easy to do that.

Thanks for the great work on this project! I've been excitingly waiting for this release.

**Reply by matallui** [reply]
Btw, here's the difference I'm seeing regarding shadows.

Before:
![image](https://user-images.githubusercontent.com/6828071/186798700-a443bd90-a895-4f7b-ba97-9b2ebfbe024d.png)

After:
![image](https://user-images.githubusercontent.com/6828071/186798718-b7b20506-7e65-4f95-a230-0a67eff7525a.png)

**Reply by matallui** [reply]
I've removed the shadows completely for now. I've noticed that even `shadow-sm` seems to be quite accented, in comparison to my old shadows. Is this the normal behavior? I feel like a small shadow should be pretty subtle like my first screenshot.

**Reply by matallui** [reply]
Sorry, maybe we can create a separate discussion on that. My initial problem is solved so I'll open a new discussion on shadows if I can't find one open.

**Reply by marklawlor** [reply]
There was a bug with the color opacity in the default thing. Version `2.0.2` should restore the shadows to how you expect

**Reply by matallui** [reply]
Just upgraded to `2.0.4` and the shadows seem to be back to normal. Thanks again!!

**Comment by danstepanov** [comment]
Closing this discussion, if this is still relevant, feel free to open a new one.
